### ISSUE TITLE
BadRefCore RefCore in packedPFCandidates (rereco job, EOYReReco2017 , Run2017F, Tau dataset)

### ISSUE THREAD

**Issue by slava77** [cms-sw/cmssw#21780]
> Hi all,
> we have been notified by computing [1] about a failure in data rereco
> affecting some job of Run2017F, Tau dataset.
> 
> It seems something happening at miniAOD level.
> 
> I report here the error message:
> 
> Fatal Exception (Exit code: 8012)
> An exception of category 'InvalidReference' occurred while
> [0] Processing Event run: 305377 lumi: 423 event: 771012315 stream: 7
> [1] Running path 'MINIAODoutput_step'
> [2] Prefetching for module PoolOutputModule/'MINIAODoutput'
> [3] Prefetching for module PATMuonSlimmer/'slimmedMuons'
> [4] Prefetching for module PATMuonSelector/'selectedPatMuons'
> [5] Prefetching for module PATMuonProducer/'patMuons'
> [6] Prefetching for module
> CITKPFIsolationSumProducerForPUPPI/'muonPUPPIIsolation'
> [7] Calling method for module
> PATPackedCandidateProducer/'packedPFCandidates'
> Exception Message:
> BadRefCore RefCore: Request to resolve a null or invalid reference to a
> product of type 'std::vector' has been detected.
> Please modify the calling code to test validity before dereferencing.
> 
> The full log of the failing job is available here [2]
> 
> Can you please have a look?
> 
> Thank you,
>     Francesco
> 
> [1]
> https://its.cern.ch/jira/browse/CMSCOMPPR-1918
> 
> [2]
> http://cms-unified.web.cern.ch/cms-unified/joblogs/vlimant_ACDC0_Run2017F-v1-Tau-17Nov2017_940_171218_115258_6053/76/DataProcessing/70059d5a-e3e3-11e7-b15a-02163e01c3a1-64-3-logArchive/cmsRun1/cmsRun1-stdout.log

@arizzi @gpetruc

**Comment by slava77** [comment]
assign reconstruction

**Comment by Dr15Jones** [comment]
Taking an educated guess, I think the exception may be coming from 
https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc#L263

On this line
https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc#L239
`PVOrig` is checked to be non-null, but on line 263 there is no such protection.

**Comment by arizzi** [comment]
Hi,
well the point is that we do not expect to have no PV. IIRC (but have no
time to debug this now) a fake, BS based, one is inserted when none is
fitted.

ciao
andrea

Il 24 dic 2017 4:35 PM, "Chris Jones" <notifications@github.com> ha scritto:

Taking an educated guess, I think the exception may be coming from
https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/plugins/
PATPackedCandidateProducer.cc#L263

On this line
https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/plugins/
PATPackedCandidateProducer.cc#L239
PVOrig is checked to be non-null, but on line 263 there is no such
protection.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/cms-sw/cmssw/issues/21780#issuecomment-353790447>, or mute
the thread

**Comment by arizzi** [comment]
ah hold on those are the  associated PV... perhaps we have a case with idx
=-1... still have no PC to debug it now...

Il 24 dic 2017 4:53 PM, "Andrea Rizzi" <andrea.rizzi@cern.ch> ha scritto:

**Comment by arizzi** [comment]
Hi,
looking again at the code, the original PV associatore code wasn't
returning -1 (because the dz association could not fail), this has been
recently changed [1] using timing info (a  phase2 development) and for
simplicity it was switched to use dz significance (because the track error
is the same for all vertices and the vertex error is not used the ranking,
this is the same as dz when no timing info is available, as long as the
track error is not zero or nan or inf).
What I suspect given the many tracking printouts in the event (but I need
to verify) is that we get a track with some bad error, hence the "min
distance" doesn't find any minimum and the algo returns -1.

Cheers,
Andrea

[1]
https://github.com/cms-sw/cmssw/commit/e02bf9d0d51d557e7d11019ee8af7d71e0f164cd

**Comment by arizzi** [comment]
I confirm that here
https://github.com/cms-sw/cmssw/blob/master/CommonTools/RecoAlgos/src/PrimaryVertexAssignment.cc#L68
we get a `-nan` from dzError().

I would suggest to fix the PVAssignment so that it doesn't return -1 (e.g. not using the significance when not needed, or checking the validity of the track error)
A fix in the PATPackedCandidate producer is also possible but I'm not sure if there are other downstream effects of this -1.. so I'd rather fix there.

@bendavid

**Comment by arizzi** [comment]
@slava77  should I add a protection and make a PR against both 10x and 9x?
@VinInn can you have a look also about why we get nan from dzError ?

**Comment by perrotta** [comment]
@arizzi : looking at the reco::Track DataFormat, dzError() can be nan if track.pt()=0 (but in that case also the dz would be a nan, which is not what you observe) or if the track covariance matrix has nan's.

Since this actually appears to happen, I'd suggest to prepare and submit the PRs which implement the protection: thank you!

**Comment by slava77** [comment]
IIUC, from the messages earlier, the design of PVAssignment is never to return -1.
It seems like it should be independent of the values in the tracks.
Let's have a fix to this.

10_0_X is clear.
For 9_4_X: is the situation such that the change will affect only the crashing events? If so, it should be straightforward as well.

**Comment by arizzi** [comment]
yes the change doesn't affect non-crashing events.

**Comment by fabozzi** [comment]
Hello, just for completeness:
The same issue has been seen in another dataset:
https://hypernews.cern.ch/HyperNews/CMS/get/prep-ops/4725.html

**Comment by slava77** [comment]
On 1/11/18 1:15 AM, fabozzi wrote:
> Hello, just for completeness:
> The same issue has been seen in another dataset:
> https://hypernews.cern.ch/HyperNews/CMS/get/prep-ops/4725.html

this was in run 305112, different from what started this thread.
So, this still remains rare, but not limited to just a single event in
one run.

**Comment by slava77** [comment]
@VinInn @makortel could you please check (or pass along to someone) to see if this nan dzerror could be captured in the track producer(s) in a relatively simple way?
This would be a better fix than the downstream kludge available in #21816 / #21817

**Comment by fabozzi** [comment]
@slava77 
> this was in run 305112, different from what started this thread. So, this still remains rare, but not limited to just a single event in one run.

exactly. I am keeping track of the failures as they are notified by comp-ops here:
https://twiki.cern.ch/twiki/bin/view/CMS/PdmVDataReprocessingEOYReReco2017#List_of_processing_problems

**Comment by VinInn** [comment]
There is NO "someone" at the moment. Tracking is not used to capture nan upstream.

**Comment by slava77** [comment]
On 1/11/18 7:43 AM, fabozzi wrote:
> @slava77 <https://github.com/slava77>
> 
>     this was in run 305112, different from what started this thread. So,
>     this still remains rare, but not limited to just a single event in
>     one run.
> 
> exactly. I am keeping track of the failures as they are notified by
> comp-ops here:
> https://twiki.cern.ch/twiki/bin/view/CMS/PdmVDataReprocessingEOYReReco2017#List_of_processing_problems

It is very nice that we have these tracked.
I noticed this yesterday.
Does this cover all the cases so far?
If so, the picture looks close to perfect :-)

**Comment by slava77** [comment]
On 1/11/18 7:43 AM, Vincenzo Innocente wrote:
> There is NO "someone" at the moment. Tracking is not used to capture nan
> upstream.
> 

I had in mind something similar to this

https://github.com/cms-sw/cmssw/blob/CMSSW_10_0_0_pre3/TrackingTools/TrackFitters/src/KFTrajectoryFitter.cc#L126

**Comment by fabozzi** [comment]
> Does this cover all the cases so far?
Yes, it should. We have always been notified so far by comp-ops before taking a decision.

**Comment by slava77** [comment]
@fabozzi 

we have a fix (upstream of the PV sorting) which gets rid of bad tracks.
It was merged in 10_1_X.
Ideally, it should still go through the pre-release relval/validation campaign before being considered for a backport.

I see that the number of events affected by this issue in data rereco in 94X is tiny.
We can still make a patch (940patch2?), but I'd make this effort only if there is a commitment to actually use it and recover these few lumisections in a way still usable for physics.
Please clarify.

**Comment by fabozzi** [comment]
@slava77 
I do not think there is plan in PPD to recover these failed lumisections in EOYReReco2017 (which would require another round of reprocessing with the patch release)

@franzoni @mklute can comment if I am wrong on this
